### PR TITLE
Parameterized cuda graph launch 3

### PR DESCRIPTION
**Pull request by galv**
This time, it is written purely in python, thanks to cuda-bindings. The previous C++ implementation was just too intrusive, since it made so many changes to CUDAGraph.cpp. I was not happy with that.

This achieves the following results on training and inference tests from the huggingface and torchbench benchmarks in a local version of torchbench:

  | accuracy | count | percent |
  |---|---:|---:|
  | pass | 191 | 73.46% |
  | eager_fail_to_run | 49 | 18.85% |
  | pass_due_to_skip | 11 | 4.23% |
  | model_fail_to_load | 4 | 1.54% |
  | fail_to_run | 2 | 0.77% |
  | eager_two_runs_differ | 1 | 0.38% |
  | fail_accuracy | 1 | 0.38% |
  | timeout | 1 | 0.38% |

The eager_fail_to_run, fail_accuracy, fail_to_run, and model_fail_to_load failures were all due to unrelated errors.

Therefore, it appears that every test case I tested runs successfully with this mode of execution. Here is the relevant way that I ran torchbench, though it is missing the relevant source code changes:

```

set -euo pipefail

workdir=trees_vs_params4

for metric in "accuracy" "performance"; do
    for method in "inference" "training"; do
        for dataset in "huggingface" "torchbench"; do
            mkdir -p $workdir/${dataset}/${method}/${metric}
            python run_benchmark.py dynamo --speedup-parameterized-cudagraphs-basic --output-directory $workdir/${dataset}/${method}/${metric} --${metric} --${method} --${dataset} 2>&1 | tee $workdir/${dataset}_params_${method}_${metric}.log
        done
    done
done

for metric in "accuracy" "performance"; do
    for method in "inference" "training"; do
        for dataset in "huggingface" "torchbench"; do
            mkdir -p $workdir/${dataset}/${method}/${metric}
            python run_benchmark.py dynamo --inductor --output-directory $workdir/${dataset}/${method}/${metric} --${method} --${metric} --${dataset} 2>&1 | tee $workdir/${dataset}_trees_${method}_${metric}.log
        done
    done
done
```

There are some missing pieces:

1. I need to do an analysis to check 0 whether the peak memory usage goes down. when using parameterized cuda graph launch instead of cuda graph trees. If this is not the case, then parameterized cuda graph launch will not be very useful in practice.

2. Today, I am replacing memcpy and memset nodes with cuda kernels. This is actually a bad thing because it means we cannot use the copy engine to perform these operations. I have a fix for this,b ut I have not merged and tested it yet.

3. I am not handling a special case where pointer parameters are writen to host memory and then copies to device memory. I also have code for this as well, but have not pushed it yet.

This code was originally written mostly in C++ at
https://github.com/pytorch/pytorch/pull/160351; it was translated to python with the help of gpt 5.5.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by galv** [comment]
The main problem with this implementation is that it actually increases memory usage relative to cuda graph trees in training workloads. This is subtle. Essentially, cached activation inputs to the backward pass die during the backward pass. Parameterized cuda graph launch uses a separate pool for inputs than for stream capture. Thus, the stream capture pool cannot reuse memory left behind by dead inputs during the backward pass in training. Ouch.

I have some experiments showing that the memory regressions, for the most part, though, are not that high compared to cuda graph trees.

https://gist.github.com/galv/42ea4ebf03d858bde131f0ea065cc535

Overall, this may still serve a purpose for running workloads that mix cudagraphs and non-cudagraphs together without serious memory regression.
